sync15-adapter/src/token.rs

```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
         Ok(Authorization(format!("Hawk {}", header)))
     }
 }
+
+// The state our TokenserverClient holds to reflect the state of the token.
+#[derive(Debug)]
+enum TokenState {
+    // We've never fetched a token.
+    NoToken,
+    // Have a token and last we checked it remained valid.
+    Token(TokenContext),
+    // We failed to fetch a token. First elt is the error, second elt is
+    // the api_endpoint we had before we failed to fetch a new token (or
+    // None if the very first attempt at fetching a token failed)
+    Failed(error::Error, Option<String>),
+    // Previously failed and told to back-off for SystemTime duration. Second
+    // elt is the api_endpoint we had before we hit the backoff error.
+    // XXX - should we roll Backoff and Failed together?
+    Backoff(SystemTime, Option<String>),
+    // api_endpoint changed - we are never going to get a token nor move out
+    // of this state.
+    NodeReassigned,
+}
+
+/// The TokenserverClient - long lived and fetches tokens on demand (eg, when
+/// first needed, or when an existing one expires.)
+#[derive(Debug)]
+pub struct TokenserverClient {
+    // The stuff needed to fetch a token.
+    base_url: String,
+    access_token: String,
+    key_id: String,
+    // Our token state (ie, whether we have a token, and if not, why not)
+    current_state: RefCell<TokenState>,
+}
+
+impl TokenserverClient {
+    pub fn new(base_url: String, access_token: String, key_id: String) -> TokenserverClient {
+        TokenserverClient {
+            base_url,
+            access_token,
+            key_id,
+            current_state: RefCell::new(TokenState::NoToken),
+        }
+    }
+
+    // Attempt to fetch a new token and return a new state reflecting that
+    // operation. If it worked a TokenState::Token state will be returned, but
+    // errors may cause other states.
+    fn fetch_token(&self, request_client: &Client, previous_endpoint: Option<String>) -> TokenState {
+        match TokenContext::new(self, request_client) {
+            Ok(tc) => {
+                // We got a new token - check that the endpoint is the same
+                // as a previous endpoint we saw (if any)
+                match previous_endpoint {
+                    Some(prev) => {
+                        if prev == tc.token.api_endpoint {
+                            TokenState::Token(tc)
+                        } else {
+                            warn!("api_endpoint changed from {} to {}", prev, tc.token.api_endpoint);
+                            TokenState::NodeReassigned
+                        }
+                    },
+                    None => {
+                        // Never had an api_endpoint in the past, so this is OK.
+                        TokenState::Token(tc)
+                    }
+                }
+            },
+            Err(e) => {
+                match error::Error::from("error!") {
```

Comment:
I think this should be matching on `e`. As it is we'll have `error::Error(error::ErrorKind::Msg("error!"))` or something like that, so there's no reason to match.